src/Zio/FileSystems/ZipArchiveFileSystem.cs

```diff
@@ -0,0 +1,1034 @@
+                }
+
+                return entry;
+            }
+            finally
+            {
+                _entriesLock.ExitWriteLock();
+            }
+        }
+
+        private string GetName(ZipArchiveEntry entry)
+        {
+            var name = entry.FullName.TrimEnd('/', '\\');
+            var index = name.LastIndexOfAny(new[] { '/', '\\' });
+            return name.Substring(index + 1);
+        }
+
+        private static string GetParent(string path)
+        {
+            path = path.TrimEnd('/', '\\');
+            var lastIndex = path.LastIndexOfAny(new[] { '/', '\\' });
+            return lastIndex == -1 ? "" : path.Substring(0, lastIndex);
+        }
+
+        private static string ConvertPathToDirectory(UPath path)
+        {
+            return ConvertPathToDirectory(path.FullName);
+        }
+
+        private static string ConvertPathToDirectory(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                return path;
+            }
+
+            return path[path.Length - 1] is DirectorySeparator ? path : path + DirectorySeparator;
+        }
+
+        private static string RemoveLeadingSlash(UPath path)
+        {
+            return path.FullName[0] is '\\' or '/' ? path.FullName.Substring(1) : path.FullName;
+        }
+
+        private static string RemoveLeadingSlash(string path)
+        {
+            return path[0] is '\\' or '/' ? path.Substring(1) : path;
+        }
+
+        private FileSystemEventDispatcher<FileSystemWatcher>? TryGetDispatcher()
+        {
+            lock (_dispatcherLock)
+            {
+                return _dispatcher;
+            }
+        }
+
+
+        private class ZipEntryStream : Stream
+        {
+            private readonly ZipArchiveEntry _entry;
+            private readonly ZipArchiveFileSystem _fileSystem;
+            private readonly Stream _streamImplementation;
+            private bool _isDisposed;
+
+            public ZipEntryStream(FileShare share, ZipArchiveFileSystem system, ZipArchiveEntry entry)
+            {
+                _entry = entry;
+                _fileSystem = system;
+                var fileShare = _fileSystem._openStreams.TryGetValue(entry, out var fileData) ? fileData.First : FileShare.ReadWrite;
+                if (fileData != null)
+                {
+                    // we only check for read share, because ZipArchive doesn't support write share
+                    if (share is not FileShare.Read and not FileShare.ReadWrite)
+                    {
+                        throw new IOException("File is already opened for reading");
+                    }
+
+                    if (fileShare is not FileShare.Read and not FileShare.ReadWrite)
+                    {
+                        throw new IOException("File is already opened for reading by another stream with non compatible share");
+                    }
+
+                    fileData.Second++;
```

Comment:
In case of multithreading, might want to use a field with `Interlocked.Increment`